_posts/2020-02-25-password-encoding-spring-security.md

```diff
@@ -0,0 +1,535 @@
+and the application supports both encodings.
+
+
+## Migrate Encoding
+If the passwords in the database are encoded by an old algorithm, then it would be better to
+migrate the passwords to another encoding. To migrate a password to another encoding we have to encode the plain text
+password. Of course, we don't have this in the database and we can't compute it without huge effort.
+Also, we can't force the users to migrate their passwords. By we can implement a
+slow gradual migration. Every time, when the user authenticates, we encode the plain password, that the user sends
+to the server and override the old one in the database. This approach is transparent to the users.
+First, we have to set `eraseCredentials` property to false in security configuration, because the Spring Security removes
+the credentials from the memory as soon as possible.  
+
+````java
+    @Autowired
+    public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception {
+        auth.authenticationProvider(daoAuthenticationProvider())
+                .eraseCredentials(false);
+    }
+````
+
+Now we can implement `ApplicationListener` for `AuthenticationSuccessEvent`.
+````java
+@Component
+public class PasswordMigration {
+
+    @Bean
+    public ApplicationListener<AuthenticationSuccessEvent> authenticationSuccessListener(
+            PasswordEncoder encoder, UserDetailsPasswordService userDetailsPasswordService) {
+        return (AuthenticationSuccessEvent event) -> {
+            Authentication authentication = event.getAuthentication();
+            User user = (User) authentication.getPrincipal();
+            String encodedPassword = user.getPassword();
+            if (encodedPassword.startsWith("{SHA-1}")) {
+                CharSequence clearTextPassword = (CharSequence) authentication.getCredentials();
+                String newPassword = encoder.encode(clearTextPassword);
+                userDetailsPasswordService.updatePassword(user, newPassword);
+            }
+            ((UsernamePasswordAuthenticationToken) authentication).eraseCredentials();
+        };
+    }
+}
+````
+
+This method is called every time when the authentication was successful. For the password update we user the 
+`UserDetailsPasswordService` we implemented before. Note, that at the end of the code we erase the credentials
+of the user.
+
+## Calculation of the Optimal Work Factor
+How to choose the suitable work factor for the password encoder? The Spring Security recommends tuning the
+password encoder to take about 1 second to verify the password. But this time depends on hardware, where the application runs.
+If the same application runs on different hardware by different customers, we can not set the best work factor to compile time.
+But we can calculate the good work factor by starting the application.
+
+````java
+ @Bean
+    public PasswordEncoder passwordEncoder() {
```

Comment:
Please include the class around this method.

Reply:
done